docs: reorganize predicted latency docs and fix broken links (#1247)
llm-d/llm-d: guides/predicted-latency-based-scheduling/README.md

# Predicted Latency-Based Scheduling

Route each inference request to the model server predicted to serve it fastest — and, optionally, only to a server predicted to meet its TTFT/TPOT SLO.

This path is for operators who want to **adopt** predicted latency-based scheduling in an existing llm-d deployment. For what the component is and how it works internally — the plugin pipeline, the ML model, scaling characteristics, the full metric list — see [architecture/advanced/latency-predictor.md](../../docs/wip-docs-new/architecture/advanced/latency-predictor.md).

## When to Pick This Path

Pick it when:

- Your workload has **high variance in prompt and completion length**, and queue depth alone is a poor proxy for true load.
- Your clients can express **per-request latency SLOs** (interactive vs. batch) and you want the gateway to enforce them.
- Static weight tuning between cache affinity and load has become **fragile** as traffic shifts.

Skip it when your pool is **heterogeneous** — mixed GPU types, model variants, or serving configurations in the same pool will produce inaccurate predictions, because the predictor assumes a single pod shape.

## Prerequisites

> - A working Inference Gateway, `InferencePool`, and at least one model server. If you don't have this, start with [getting-started/quickstart.md](../../docs/wip-docs-new/getting-started/quickstart.md).

Reviewer: pls update as per the new optimized-baseline guide.

Author: done

Reviewer: pls remove this prereq and apply the same pattern to setting up the gateway mode that we used in optimized-baseline: https://github.com/llm-d/llm-d/blob/main/guides/optimized-baseline/README.md?plain=1#L85

Author: i still have a prerequisite just like in optimized-baseline but the pattern in mirrored

## Deploy

Enable the predictor by setting `inferenceExtension.latencyPredictor.enabled=true` when installing or upgrading the chart. This wires up the full plugin pipeline automatically — prediction, cache affinity filtering, SLO tier gating, scoring, admission control, and weighted endpoint selection. The SLO plugins stay idle until a request actually carries SLO headers, so it's safe to enable even for traffic that doesn't use SLOs yet.

```bash
helm install <release> . \
  --set inferencePool.modelServers.matchLabels.app=<your-model-label> \
  --set inferenceExtension.latencyPredictor.enabled=true \
  --set inferenceExtension.monitoring.prometheus.enabled=true \
  --set provider.name=gke \
  -f values.yaml
```

### If You Want SLO Enforcement: Turn On Streaming Mode

The `predicted-latency-producer` plugin has a `streamingMode` parameter. Its default (`false`) trains on end-to-end latency and does not train TPOT — fine for routing-only use, but incompatible with `x-slo-ttft-ms` / `x-slo-tpot-ms`. **Set `streamingMode: true` whenever you use SLO headers**, and make sure every request is actually streamed.

> Override it via `inferenceExtension.pluginsCustomConfig` in `values.yaml`:

Reviewer: Provide an override to the chart (see https://github.com/llm-d/llm-d/blob/main/guides/optimized-baseline/scheduler/optimized-baseline.values.yaml as an example) that the user applies using -f flag.

- A working Inference Gateway, `InferencePool`, and at least one model server. If you don't have this, start with [getting-started/quickstart.md](../../docs/wip-docs-new/getting-started/quickstart.md).
- Helm access to the chart used to deploy the EPP and inference pool.
- A **homogeneous** pool — same GPU type, same model weights, same serving config across every pod.
- If you plan to use SLO headers: a **fully streaming** workload (`"stream": true` on every request).

## Deploy

Enable the predictor by setting `inferenceExtension.latencyPredictor.enabled=true` when installing or upgrading the chart. This wires up the full plugin pipeline automatically — prediction, cache affinity filtering, SLO tier gating, scoring, admission control, and weighted endpoint selection. The SLO plugins stay idle until a request actually carries SLO headers, so it's safe to enable even for traffic that doesn't use SLOs yet.

Two ready-to-use values files ship with this guide:

| File | When to use |
|---|---|
| [`scheduler/predicted-latency.values.yaml`](./scheduler/predicted-latency.values.yaml) | Default — predictor trains on end-to-end latency. Routing-only, no SLO header support. |

> | [`scheduler/predicted-latency-slo.values.yaml`](./scheduler/predicted-latency-slo.values.yaml) | SLO-aware — sets `streamingMode: true` on `predicted-latency-producer`. Required for `x-slo-ttft-ms` / `x-slo-tpot-ms`. Every request must be sent with `"stream": true`. |

Reviewer: ```suggestion
| [`scheduler/predicted-latency-slo.values.yaml`](./scheduler/predicted-latency-slo.values.yaml) | SLO-aware — Assumes `x-slo-ttft-ms` / `x-slo-tpot-ms` are set on requests. Every request must be sent with `"stream": true`. |
```

> Both files pin the EPP image to `ghcr.io/llm-d/llm-d-inference-scheduler:v0.8.0-rc.2` (the GIE-v1.5.0-compatible build) and target model server pods labeled `llm-d.ai/guide=optimized-baseline`. Edit the `inferencePool.modelServers.matchLabels` block in the file you pick to retarget your own pods.

Reviewer: ```suggestion
Both target model server pods labeled `llm-d.ai/guide=optimized-baseline` since in the next step we will simply reuse the model server manifests from the [optimized-baseline guide](../optimized-baseline).
```

```bash
# Default (routing-only)
helm install predicted-latency-scheduler \
  oci://registry.k8s.io/gateway-api-inference-extension/charts/standalone \
  --version v1.5.0 \
  -f guides/predicted-latency-based-scheduling/scheduler/predicted-latency.values.yaml
```

```bash
# SLO-aware (streamingMode: true)
helm install predicted-latency-scheduler \
  oci://registry.k8s.io/gateway-api-inference-extension/charts/standalone \
  --version v1.5.0 \
  -f guides/predicted-latency-based-scheduling/scheduler/predicted-latency-slo.values.yaml
```

> [!NOTE]

> > Do not layer the shared `guides/recipes/scheduler/base.values.yaml` on top of either file — it is pinned to the older v1.4.0 image-key schema (`hub`/`name`), which v1.5.0 renamed to `registry`/`repository`.

Reviewer: pls remove this note, the step above is clear on how the user should be deploying this.

### Streaming Mode and SLO Headers

> The `predicted-latency-producer` plugin has a `streamingMode` parameter. Its default (`false`, used by `predicted-latency.values.yaml`) trains on end-to-end latency and does not train TPOT — fine for routing-only use, but incompatible with `x-slo-ttft-ms` / `x-slo-tpot-ms`. **Use `predicted-latency-slo.values.yaml` whenever you send SLO headers**, and make sure every request is actually streamed.

Reviewer: this is a repeat of the above, recommend removing.

> ### Scoring Strategy

Reviewer: pls remove, this is discussed in the predicted-latency architectural guide, not need to repeat here.

    --set experimentalHttpRoute.inferenceGatewayName=llm-d-inference-gateway \
    -n ${NAMESPACE} --version ${GAIE_VERSION}
```

</details>

### 2. Deploy the Model Server

Reuse the model server manifests from the [optimized-baseline guide](../optimized-baseline/) — the values files above already select pods labeled `llm-d.ai/guide=optimized-baseline`.

```bash
kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/gpu/vllm/
```

## Send Requests

Once enabled, latency-based scheduling works on every request — no header changes needed. The proxy picks the endpoint with the lowest predicted latency.

To opt an individual request into SLO-aware routing, add one or both headers:

- `x-slo-ttft-ms` — Time-to-first-token SLO in milliseconds.
- `x-slo-tpot-ms` — Time-per-output-token SLO in milliseconds.

### 1. Get the IP of the Proxy

**Standalone Mode**

```bash
export IP=$(kubectl get service ${GUIDE_NAME}-epp -n ${NAMESPACE} -o jsonpath='{.spec.clusterIP}')
```

<details>
<summary> <b>Gateway Mode</b> </summary>

```bash
export IP=$(kubectl get gateway llm-d-inference-gateway -n ${NAMESPACE} -o jsonpath='{.status.addresses[0].value}')
```
</details>

> ### 2. Send a Test Request

Reviewer: pls mirror https://github.com/llm-d/llm-d/tree/main/guides/optimized-baseline#2-send-test-requests; we need a step to open a temporary interactive shell inside the cluster to reach the private IP.

    -f guides/${GUIDE_NAME}/scheduler/predicted-latency.values.yaml \
    -n ${NAMESPACE} --version ${GAIE_VERSION}
```

For SLO-aware scheduling, swap the values file: `-f guides/${GUIDE_NAME}/scheduler/predicted-latency-slo.values.yaml`.

<details>
<summary><h4>Gateway Mode</h4></summary>

To use a Kubernetes Gateway managed proxy rather than the standalone version, follow these steps instead of applying the previous Helm chart:

1. *Deploy a Kubernetes Gateway* by following one of [the gateway guides](../prereq/gateways).
2. *Deploy the inference scheduler and an HTTPRoute* that connects it to the Gateway as follows:

```bash
export PROVIDER_NAME=gke # options: none, gke, agentgateway, istio
helm install ${GUIDE_NAME} \
    oci://registry.k8s.io/gateway-api-inference-extension/charts/inferencepool \
    -f guides/${GUIDE_NAME}/scheduler/predicted-latency.values.yaml \
    --set provider.name=${PROVIDER_NAME} \
    --set experimentalHttpRoute.enabled=true \
    --set experimentalHttpRoute.inferenceGatewayName=llm-d-inference-gateway \
    -n ${NAMESPACE} --version ${GAIE_VERSION}
```

</details>

### 2. Deploy the Model Server

> This guide reuses the model server manifests from the optimized-baseline guide (the values files above already select pods labeled `llm-d.ai/guide=optimized-baseline`). Follow [optimized-baseline → Deploy the Model Server](../optimized-baseline/README.md#2-deploy-the-model-server) to apply the kustomize overlay for your backend (NVIDIA GPU, AMD GPU, Intel XPU, Gaudi, TPU, or CPU).

Reviewer: I think we should have have an explicit command here for the default gpu deployment:

kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/gpu/vllm/

Author: done
